File: scripts/governance/verify_integration_pairing.py
#!/usr/bin/env python3
"""Buddy System — new Python source files need integration tests"""

import subprocess
import sys
from pathlib import Path


def get_new_source_files() -> list[Path]:
    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only", "--diff-filter=A"],
            capture_output=True,
            text=True,
        )
        return [
            Path(f)
            for f in result.stdout.strip().split("\n")
            if f.startswith("app/") and f.endswith(".py") and Path(f).exists()
        ]
    except Exception:
        return []


new_files = get_new_source_files()
if not new_files:
    print("✓ Integration pairing: no new source files added")
    sys.exit(0)

missing = []
for src in new_files:
    stem = src.stem
    integration_file = Path(f"tests/integration/test_{stem}_integration.py")
    if not integration_file.exists():
        missing.append(f"  {src} → missing {integration_file}")

if missing:
    print(
        "✗ Buddy System: new source files missing integration tests:", file=sys.stderr
    )
    for m in missing:
        print(m, file=sys.stderr)
    sys.exit(1)

print(f"✓ Integration pairing: all {len(new_files)} new file(s) have integration tests")

File: scripts/governance/verify-integration-pairing.js
#!/usr/bin/env node
/**
 * Buddy System — every source file must have an integration test counterpart.
 * Checks staged .ts files in src/ for corresponding tests/integration/*.test.ts
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

function getStagedSourceFiles() {
  try {
    const output = execSync('git diff --cached --name-only --diff-filter=A', { encoding: 'utf8' });
    return output.trim().split('\n')
      .filter(f => f.startsWith('src/') && f.endsWith('.ts') && !f.includes('__tests__') && !f.endsWith('.d.ts'));
  } catch { return []; }
}

const newSourceFiles = getStagedSourceFiles();

if (newSourceFiles.length === 0) {
  console.log('✓ Integration pairing: no new source files added');
  process.exit(0);
}

const missing = [];

for (const srcFile of newSourceFiles) {
  const base = path.basename(srcFile, '.ts');
  const integrationFile = `tests/integration/${base}.integration.test.ts`;
  if (!fs.existsSync(integrationFile)) {
    missing.push(`  ${srcFile} → missing ${integrationFile}`);
  }
}

if (missing.length > 0) {
  console.error('✗ Buddy System: new source files missing integration tests:');
  missing.forEach(m => console.error(m));
  console.error('  Add an integration test before committing new source files.');
  process.exit(1);
}

console.log(`✓ Integration pairing: all ${newSourceFiles.length} new file(s) have integration tests`);
process.exit(0);
